protocol/discord/types.go

```diff
@@ -57,10 +113,150 @@ type ImportError struct {
 	// Non-critical errors are the ones that would not prevent the imported
 	// community from functioning. For example, if the channel data to be imported
 	// has no messages, or is not parsable.
-	Code    ErrorType `json:"code"`
-	Message string    `json:"message"`
+	Code     ErrorCodeType `json:"code"`
+	Message  string        `json:"message"`
+	TaskInfo string        `json:"taskInfo"`
 }
 
 func (d ImportError) Error() string {
 	return fmt.Sprintf("%d: %s", d.Code, d.Message)
 }
+
+func Error(message string) *ImportError {
+	return &ImportError{
+		Message: message,
+		Code:    ErrorType,
+	}
+}
+
+func Warning(message string) *ImportError {
+	return &ImportError{
+		Message: message,
+		Code:    WarningType,
+	}
+}
+
+type ImportTaskProgress struct {
+	Type          string         `json:"type"`
+	Progress      float32        `json:"progress"`
+	Errors        []*ImportError `json:"errors"`
+	Stopped       bool           `json:"stopped"`
+	ErrorsCount   uint           `json:"errorsCount"`
+	WarningsCount uint           `json:"warningsCount"`
+	State         string         `json:"state"`
+}
+
+type ImportTasks map[ImportTask]*ImportTaskProgress
+
+type ImportProgress struct {
+	CommunityID   string                `json:"communityId,omitempty"`
+	CommunityName string                `json:"communityName"`
+	Tasks         []*ImportTaskProgress `json:"tasks"`
+	Progress      float32               `json:"progress"`
+	ErrorsCount   uint                  `json:"errorsCount"`
+	WarningsCount uint                  `json:"warningsCount"`
+	Stopped       bool                  `json:"stopped"`
+	m             sync.Mutex
+}
+
+func (progress *ImportProgress) Init(tasks []ImportTask) {
+	progress.Progress = 0
+	progress.Tasks = make([]*ImportTaskProgress, 0)
+	for _, task := range tasks {
+		progress.Tasks = append(progress.Tasks, &ImportTaskProgress{
+			Type:          task.String(),
+			Progress:      0,
+			Errors:        []*ImportError{},
+			Stopped:       false,
+			ErrorsCount:   0,
+			WarningsCount: 0,
+			State:         TaskStateInitialized.String(),
+		})
+	}
+	progress.ErrorsCount = 0
+	progress.WarningsCount = 0
+	progress.Stopped = false
+}
+
+func (progress *ImportProgress) Stop() {
+	progress.Stopped = true
+}
+
+func (progress *ImportProgress) AddTaskError(task ImportTask, err *ImportError) {
+	progress.m.Lock()
+	for i, t := range progress.Tasks {
+		if t.Type == task.String() {
+			errorsAndWarningsCount := progress.Tasks[i].ErrorsCount + progress.Tasks[i].WarningsCount
+			if errorsAndWarningsCount < MaxTaskErrorItemsCount {
+				errors := progress.Tasks[i].Errors
+				progress.Tasks[i].Errors = append(errors, err)
+			}
```

Comment:
While I've built the import tool, I ran into perf issues the bigger the `errors` list of an individual task gets.
This is because, when for example the "download assets" task keeps timing out when requesting asset payloads, there's many errors added to the task (this can be hundreds or even thousands).

When the progress is then published to clients, they have a lot of data to parse in a single signal.
And since so many errors/issues aren't really usable to a user, we decided to max cap visible errors per task to 3

If there's more than 3 errors per task we only increase the counter and let clients/UIs decide how to handle this.